net/autossh/src/opnsense/mvc/app/controllers/VerbNetworks/Autossh/Api/ServiceController.php

```diff
@@ -0,0 +1,228 @@
+            $backend_result = $this->backendStandardAction('start_tunnel', $this->request->getPost('id'));
+            if (false !== strpos($backend_result, 'Starting autossh tunnel ')) {
+                $response = array('status'=>'success', 'message' => $backend_result);
+            } else {
+                $response['message'] = 'Backend request failed';
+            }
+        }
+        return $response;
+    }
+
+    public function restartAction()
+    {
+        $response = array('status'=>'fail', 'message' => 'Invalid request');
+        if ($this->request->isPost() && $this->request->hasPost('id')) {
+            $backend_result = $this->backendStandardAction('restart_tunnel', $this->request->getPost('id'));
+            if (false !== strpos($backend_result, 'Starting autossh tunnel ')) {
+                $response = array('status'=>'success', 'message' => $backend_result);
+            } else {
+                $response['message'] = 'Backend request failed';
+            }
+        }
+        return $response;
+    }
+    
+    public function stopAction()
+    {
+        $response = array('status'=>'fail', 'message' => 'Invalid request');
+        if ($this->request->isPost() && $this->request->hasPost('id')) {
+            $backend_result = $this->backendStandardAction('stop_tunnel', $this->request->getPost('id'));
+            if (false !== strpos($backend_result, 'Stopping autossh tunnel ')) {
+                $response = array('status'=>'success', 'message' => $backend_result);
+            } else {
+                $response['message'] = 'Backend request failed';
+            }
+        }
+        return $response;
+    }
+    
+    public function connectionStatusAction()
+    {
+        $response = array(
+            'current' => 1,
+            'rowCount' => null,
+            'rows' => array(),
+            'total' => null
+        );
+        
+        if ($this->request->isGet()) {
+            $model = new Autossh();
+            $grid = new UIModelGrid($model->tunnels->tunnel);
+        
+            $grid_data = $grid->fetchBindRequest($this->request, array(
+                'enabled', 'user', 'hostname', 'port', 'bind_interface', 'ssh_key',
+                'local_forward', 'remote_forward', 'dynamic_forward', 'tunnel_device'
+            ), 'hostname');
+            
+            foreach ($grid_data['rows'] as $tunnel) {
+                $connection = $tunnel['user'].'@'.$tunnel['hostname'];
+                if (!empty($tunnel['port'])) {
+                    $connection .= ':'.$tunnel['port'];
+                }
+                
+                $forward_data = array(
+                    'local' => htmlspecialchars($tunnel['local_forward']),
+                    'dynamic' => htmlspecialchars($tunnel['dynamic_forward']),
+                    'remote' => htmlspecialchars($tunnel['remote_forward']),
+                    'tunnel' => htmlspecialchars($tunnel['tunnel_device'])
```

Comment:
This is an api, not a html renderer. No escaping here. Especially because of this ugly hack:
opnsense/core#2672

Escaping must be done in the viewer - which is the js code of the frontend.

Reply:
Yes, I'm aware of https://github.com/opnsense/core/issues/2672 if you take a look at my  `TunnelsController` I specifically have an `afterExecuteRoute()` function to be able to step outside it in a very limited case - even with me implementing this I guess it did not occur to me that using `htmlspecialchars()` here is in-fact double-escaping things - either way, resolved